lndclient/macaroon_pouch.go

```diff
@@ -0,0 +1,98 @@
+package lndclient
+
+import (
+	"context"
+	"encoding/hex"
+	"io/ioutil"
+	"path/filepath"
+
+	"google.golang.org/grpc/metadata"
+)
+
+// serializedMacaroon is a type that represents a hex-encoded macaroon. We'll
+// use this primarily vs the raw binary format as the gRPC metadata feature
+// requires that all keys and values be strings.
+type serializedMacaroon string
+
+// newSerializedMacaroon reads a new serializedMacaroon from that target
+// macaroon path. If the file can't be found, then an error is returned.
+func newSerializedMacaroon(macaroonPath string) (serializedMacaroon, error) {
+	macBytes, err := ioutil.ReadFile(macaroonPath)
+	if err != nil {
+		return "", err
+	}
+
+	return serializedMacaroon(hex.EncodeToString(macBytes)), nil
+}
+
+// WithMacaroonAuth modifies the passed context to include the macaroon KV
+// metadata of the target macaroon. This method can be used to add the macaroon
+// at call time, rather than when the connection to the gRPC server is created.
+func (s serializedMacaroon) WithMacaroonAuth(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "macaroon", string(s))
+}
```

Comment:
What is the impact of this in terms of network traffic and performance (is the macaroon validated in `lnd` for each call?) compared to opening one tcp connection per sub server and using connection level macaroons?

Reply:
It's the exact same. There's a special header that's used to tack on arbitrary "metadata" which are KV pairs: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests (search for "metadata").

Reply:
That's nice. Then the chosen solution is definitely the best by having just a single tcp connection.